### ISSUE TITLE
[Bug] TestBenchServingDenseTp4 not run successfully on main branch

### ISSUE THREAD

**Issue by pengchengneo** [sgl-project/sglang-jax#1044]
### Checklist

- [ ] 1. I have searched related issues but cannot get the expected help.
- [ ] 2. The bug has not been fixed in the latest version.
- [ ] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [ ] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [ ] 5. Please use English, otherwise it will be closed.

### Describe the bug

<img width="2132" height="864" alt="Image" src="https://github.com/user-attachments/assets/a209ff00-da7b-41be-91ef-46eb8575cec3" />

### Reproduction

in ci e2e test, the performance will up and down fluctuation 0.2% for output throupht

### Environment

tpu v6e

**Comment by JamesBrianD** [comment]
### Root cause: `bq_sz=1` default in DECODE sub-kernel

Reproduced on tpu-v6e-4 (Qwen3-8B, TP=4, 256 concurrent decoders, JAX 0.8.1).

#### Evidence

Same commit `febb350c` (head of #934), single run each:

| # | Configuration | output_throughput (tok/s) | Δ |
|---|---|---|---|
| A | PR #934 as-is (`distribution=[N,N,N]` → DECODE sub-kernel) | **9985.12** | baseline |
| B | Revert distribution one-liner (`[0,0,N]` → MIXED sub-kernel) | **11802.58** | **+18.2%** |
| C | Keep DECODE routing, only raise `bq_csz=32` | 9955.50 | −0.3% (no effect) |
| D | Keep DECODE routing, copy MIXED's full block-size formula into the DECODE branch | **11684.09** | **+17.0%** |

Historical CI matches: PR #930 ran 11484.65 tok/s; PR #934 9965.90 tok/s. Recent main runs land at 9779–9839 tok/s, tripping the `assertGreater(9866)` threshold (e.g. runs 25787533738, 25713412493, 25668305532).

#### Mechanism

PR #934 routes decode-only batches into `RpaCase.DECODE`:
```python
# python/sgl_jax/srt/layers/attention/flashattention_backend.py:153
distribution = np.array([num_seqs, num_seqs, num_seqs], dtype=np.int32)  # was [0, 0, num_seqs]
```
which causes `_prepare_block_sizes` to call `get_default_block_sizes(case=RpaCase.DECODE)`:
```python
# python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention_v3.py:1497
case 5 | 6:
    if case == RpaCase.DECODE:
        bq_sz   = 1
        bq_csz  = 1
        ...
    else:                                          # MIXED
        bq_sz   = min(MAX_BQ_SZ, max_q // 2)        # = 32
        bq_csz  = min(MAX_BQ_SZ, min(512 // num_q_heads_per_kv_head, max_q))  # = 32
```
And inside `process()`:
```python
# ragged_paged_attention_v3.py:875-883
if static_q_len is None:                # MIXED path
    actual_bq_sz = bq_sz
else:                                   # DECODE path, static_q_len=1
    actual_bq_sz = min(bq_sz, static_q_len)   # = min(N, 1) = 1
actual_bq_csz = min(bq_csz, actual_bq_sz)     # forced back to 1
```

Two consequences:
1. Tweaking only `bq_csz` is silently clamped to 1 because `static_q_len=1` forces `actual_bq_sz=1` (this is why experiment C had no effect).
2. With `bq_sz=1` the kernel runs `num_seqs=256` grid steps, paying KV-pipeline cost per step → flat ~10220 tok/s ceiling. With `bq_sz=32` it runs `⌈256/32⌉=8` steps and amortizes KV reload across 32 query rows. `bq_sz=N` doesn't change correctness — it processes N sequences' single token in parallel along the sequence axis, not multiple tokens for one sequence.

#### Why this is a kernel-tuning issue, not a routing bug

The DECODE routing in #934 is correct and matches what tpu-inference's vLLM serving layer does (`runner/persistent_batch_manager.py` and `runner/tpu_runner.py` both build `[num_decode, num_decode, num_reqs]`). The same `bq_sz=1, bq_csz=1` defaults exist verbatim in upstream `tpu_inference/kernels/ragged_paged_attention/v3/kernel.py:1513-1517` (introduced by tpu-inference commit b2e416f0, "[RPA3] Separate kernel to 3 calls and many optimizations"). The kernel default is presumably tuned for low-latency single-request decode, not high-concurrency throughput decode.

#### Suggested fix

Make `get_default_block_sizes(case=DECODE)` batch-aware so `bq_sz` scales with `max_num_seqs`:
```python
if case == RpaCase.DECODE:
    bq_sz   = min(MAX_BQ_SZ, max(1, max_num_seqs))   # was 1
    bkv_sz  = min(min_bkv_sz_to_peak, max_kv) if sliding_window is None else sliding_window
    bq_csz  = min(MAX_BQ_SZ, min(512 // num_q_heads_per_kv_head, max_num_seqs))
    bkv_csz = min(512, align_to(max_kv // 2, page_size)) if max_kv > 1 else page_size
```
Experiment D (full MIXED formula in the DECODE branch) restores throughput to 11684 tok/s with no change to `distribution` routing.

#### Reproduction

```bash
# Pod: TPU v6e 2x2, image us-docker.pkg.dev/cloud-tpu-images/jax-ai-image/tpu:jax0.8.1-rev1
git clone https://github.com/sgl-project/sglang-jax.git && cd sglang-jax
git checkout febb350cd8667a860ba49fffd644d8855d831668   # PR #934 head
python3.12 -m venv .venv && source .venv/bin/activate
pip install uv && uv pip install -e "python[all]"

SGLANG_JAX_IS_IN_CI=true SGLANG_JAX_MODEL_CACHE=<path-with-Qwen/Qwen3-8B> \
JAX_COMPILATION_CACHE_DIR=/tmp/jax_compilation_cache \
python -u test/srt/test_bench_serving_dense_tp_4.py \
    TestBenchServingDenseTp4.test_output_throughput_default_tp_4
```
Add a one-line `print` of `bq_sz/bkv_sz/bq_csz/bkv_csz` in `get_default_block_sizes` to see the values picked by each precompile step.

**Comment by JamesBrianD** [comment]
### Update: isolated the regression to `bkv_sz`, not `bq_sz`

Ran the full controlled-variable sweep on the same `febb350c` commit (PR #934 head), TPU v6e-4, single run each. For each variant only the named field(s) of the `case == RpaCase.DECODE` branch in `get_default_block_sizes` (case 5|6) were swapped to MIXED's formula:

| Variant | bq_sz | bkv_sz | bq_csz | bkv_csz | output_throughput | Δ |
|---|---|---|---|---|---|---|
| A baseline (PR #934 as-is) | 1 | 16384 | 1 | 16384 | **9969.67** | — |
| E only `bq_sz` → MIXED | 32 | 16384 | 1 | 16384 | **11763.81** | +18.0% |
| F only `bkv_sz` → MIXED | 1 | 1024 | 1 | 1024 (= bkv_sz) | **11919.06** | **+19.6%** |
| G only `bkv_csz` → MIXED | 1 | 16384 | 1 | 512 | 9988.57 | +0.2% (no effect) |
| H `bq_sz` + `bkv_sz` | 32 | 1024 | 1 | 1024 | 11469.34 | +15.0% |
| I `bkv_sz` + `bkv_csz` | 1 | 1024 | 1 | 512 | 11217.62 | +12.5% |
| D full MIXED formula | 32 | 1024 | 1 | 512 | 11817.76 | +18.5% |

(Values computed for Qwen3-8B/TP=4: `num_q_heads_per_kv_head=4`, `MAX_BQ_SZ=32`, `max_q≈256`, `max_kv≈40832`, `page_size=256`, `min_bkv_sz_to_peak≈16384`. `bq_csz` and `bkv_csz` get clamped to `actual_bq_sz=min(_, static_q_len=1)=1` inside the kernel, so changing them alone has no observable effect.)

#### Real root cause: `bkv_sz=16384` (i.e. `min_bkv_sz_to_peak`) is too large for high-concurrency decode

The DECODE branch defaults to:
```python
bkv_sz = min(min_bkv_sz_to_peak, max_kv)   # ≈ min(16K tokens, 40K tokens) = 16K
bkv_csz = bkv_sz
```
The 16K-token target is sized to "saturate MXU/HBM bandwidth from a single sequence's KV stream" — appropriate when there's only one request in flight and you need a fat KV tile to amortize the pipeline. But with 256 concurrent decoders the KV pipeline is already amortized across sequences, so the oversized per-seq KV tile just wastes HBM bandwidth on padding/over-fetch. Variant F (only `bkv_sz` shrunk to 1024) recovers throughput by itself with a one-line change.

`bq_sz` is a parallel issue (variant E shows it also recovers ~18%), but the two changes don't compose — variant H (both) is *worse* than either alone, and full-replacement D is no better than F. They're substitutes, not additive: once the per-step VMEM/bandwidth pressure is relieved by either lever, the other adds overhead.

#### Why we can't just key off `seq_lens`

`bkv_sz` is a static Pallas tile parameter — it bakes into buffer shapes, scratch sizes, and inner-loop unrolling, so it can't be derived from runtime `kv_lens`. The current `min(min_bkv_sz_to_peak, max_kv)` already adapts to the compile-time `max_kv = pages_per_seq × page_size`; what's missing is adapting to `max_num_seqs`.

#### Suggested fix (one-liner in `get_default_block_sizes`)

Make `bkv_sz` aware that the KV pipeline is amortized across concurrent sequences:
```python
case 5 | 6:
    if case == RpaCase.DECODE:
        bq_sz = 1
        # When max_num_seqs is large, the KV pipeline is amortized across
        # sequences, so a smaller per-seq KV tile is sufficient.
        bkv_sz_target = max(page_size, min_bkv_sz_to_peak // max(1, max_num_seqs // 8))
        bkv_sz = min(bkv_sz_target, max_kv) if sliding_window is None else sliding_window
        bq_csz = 1
        bkv_csz = bkv_sz
```
For `max_num_seqs=1` (latency-critical batch-1) this keeps the original 16K tile; for `max_num_seqs=256` it shrinks to ~512, in the same ballpark as MIXED's 1024. Same change should mirror to case 7 (v7).

A simpler alternative is to just hardcode `bkv_sz = min(1024, max_kv)` (variant F). It gives the highest throughput in our sweep (11919 tok/s) but trades away the deep-pipeline tuning for batch=1 / long-context latency, which we haven't benchmarked.

The same `bq_sz=1, bkv_sz=min(min_bkv_sz_to_peak, max_kv)` defaults exist verbatim in upstream `tpu_inference/kernels/ragged_paged_attention/v3/kernel.py:1513-1517` (commit b2e416f0), so the fix likely benefits upstream too.

**Comment by JamesBrianD** [comment]
### Recommendation: precompile decode along both `(bs, max_seq_lens)` dimensions

Today the control plane pads every decode batch's per-seq KV allocation to the model's full `context_len` (`tp_worker.py:180` → `compilation_manager.py:94` → `cache_loc_buckets[bs] = bs * pages_per_req`, where `pages_per_req` is sized from `max_req_len ≈ context_len`). RPA v3's default heuristic `bkv_sz = min(min_bkv_sz_to_peak, max_kv)` then sees a `max_kv` close to `context_len` for every batch, regardless of the actual sequence lengths in flight.

This is what produces the regression on `TestBenchServingDenseTp4`: Qwen3-8B has `context_len ≈ 40K`, so `max_kv ≈ 40832` and `bkv_sz` snaps to **16384** — even though the test only ever holds ≈1K-token sequences. The DMA tile is ~16× larger than the working set, wasting HBM bandwidth.

Verified with `--context-length 2048` on the same PR #934 commit (no kernel changes, no `distribution` revert): throughput jumps from 9985 → **14043 tok/s** (+41% over PR #934, +22% over PR #930). The kernel picks `bkv_sz = min(16384, 2048) = 2048` automatically because `max_kv` drops below the heuristic ceiling.

For production this implies the precompile manager should bucket along **both `bs` and `max_seq_lens`** rather than only `bs`. Each (bs, max_seq_lens) bucket would precompile a forward with `pages_per_req` sized to the bucket's seq-len cap, letting RPA's existing `min(min_bkv_sz_to_peak, max_kv)` heuristic adapt to the actual workload — short-prompt high-concurrency batches get small `bkv_sz`, long-context batches get large `bkv_sz`.

Trade-offs:
- compile cache count grows from N (bs buckets) to N×M (M = seq-len buckets); cache size and warmup time scale accordingly
- bucketing thresholds need tuning to avoid sparse buckets / frequent recompiles
- dispatch logic needs to round each request up to its (bs, max_seq_lens) bucket

Likely worth it for serving workloads with bimodal length distributions (e.g. short chat + long-context coexisting). For the immediate CI fix, lowering `--context-length` in `TestBenchServingDenseTp4` is a one-line workaround, but the structural issue is the single-dimension bucketing.
